core/store/src/contract.rs

```diff
@@ -24,61 +24,145 @@ pub struct ContractStorage {
+    ///
+    /// This is called when preparing (eg. compiling) the contract for execution optimistically, and
+    /// the function call for the returned contract may not be included in the currently applied chunk.
+    /// Thus, the `record_call` method should also be called to indicate that the calling the contract
+    /// is actually included in the chunk application.
     pub fn get(&self, code_hash: CryptoHash) -> Option<ContractCode> {
         {
-            let guard = self.uncommitted_deploys.lock().expect("no panics");
-            if let Some(v) = guard.get(&code_hash) {
-                return Some(ContractCode::new(v.code().to_vec(), Some(code_hash)));
+            let tracker = self.tracker.lock().expect("no panics");
+            if let Some(contract) = tracker.get(code_hash.into()) {
+                return Some(ContractCode::new(contract.code().to_vec(), Some(code_hash)));
             }
         }
 
-        let contract_code = match self.storage.retrieve_raw_bytes(&code_hash) {
+        match self.storage.retrieve_raw_bytes(&code_hash) {
             Ok(raw_code) => Some(ContractCode::new(raw_code.to_vec(), Some(code_hash))),
             Err(_) => None,
-        };
-
-        if contract_code.is_some() {
-            let mut guard = self.storage_reads.lock().expect("no panics");
-            guard.as_mut().expect("must not be called after finalize").insert(CodeHash(code_hash));
         }
+    }
+
+    /// Records an uncommitted call to a contract by code-hash.
+    ///
+    /// This is used to capture the contracts that are called when applying a chunk.
+    /// Calling `rollback` clears this recording.
+    pub fn record_call(&self, code_hash: CryptoHash) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.call(code_hash.into());
+    }
+
+    /// Stores the contract code as an uncommitted deploy.
+    ///
+    /// Subsequent calls to `get` will return the code that was stored here.
+    /// Calling `rollback` clears this uncommitted deploy.
+    pub fn record_deploy(&self, code: ContractCode) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.deploy(code);
+    }
 
-        contract_code
+    /// Commits the uncommitted recording of contract calls and deployments.
+    ///
+    /// This adds the uncommitted calls and deployments to the committed list and clears the uncommitted list.
+    /// Note that there can be multiple calls to `commit`.
+    pub(crate) fn commit(&mut self) {
+        let mut tracker = self.tracker.lock().expect("no panics");
+        tracker.commit();
     }
 
-    pub fn store(&self, code: ContractCode) {
-        let mut guard = self.uncommitted_deploys.lock().expect("no panics");
-        guard.insert(*code.hash(), code);
+    /// Rolls back the uncommitted recording of contract calls and deployments.
+    pub(crate) fn clear_uncommitted(&mut self) {
+        let mut tracker = self.tracker.lock().expect("no panics");
```

Comment:
since `self` is `&mut Self`, you can use `Mutex::get_mut` to avoid locking altogether.

Reply:
Ah, in that case I would keep the code the way it was before (i.e. without an `Option`) and instead of `iter().cloned()`ing the relevant data structures I would either `mem::take()` or `drain()` them out.

Reply:
Yes, I will send it as a separate follow-up PR finalize changes in this PR. Adding a TODO for that.

Reply:
never mind, I will do that in this PR.